packages/remix-server-runtime/serialize.ts

```diff
@@ -1,16 +1,41 @@
 import type { Jsonify } from "./jsonify";
 import type { TypedDeferredData, TypedResponse } from "./responses";
+import type {
+  ClientActionFunctionArgs,
+  ClientLoaderFunctionArgs,
+} from "./routeModules";
 import { expectType } from "./typecheck";
 import { type Expect, type Equal } from "./typecheck";
 
+// prettier-ignore
+/**
+ * Infer JSON serialized data type returned by a loader or action, while
+ * avoiding deserialization if the input type if it's a clientLoader or
+ * clientAction that returns a non-Response
+ *
+ * For example:
+ * `type LoaderData = SerializeFrom<typeof loader>`
+ */
+export type SerializeFrom<T> = (
+  T extends (...args: any[]) => infer Output
+    ? Output extends Promise<TypedResponse>
+      ? SerializeFromImpl<T>
+      : Parameters<T> extends [ClientLoaderFunctionArgs | ClientActionFunctionArgs]
+        ? Output extends Promise<TypedDeferredData<infer D>>
+          ? D // client defer() is just the raw data
+          : Awaited<Output> // client naked object is just the object
+        : SerializeFromImpl<T>
+    : SerializeFromImpl<T>
+);
```

Comment:
Trying out my own suggestion here (https://twitter.com/pcattori/status/1734712411700023443):

```suggestion
export type SerializeFrom<T> = (
  T extends (...args: any[]) => infer Output ?
    Output extends Promise<TypedResponse> ? SerializeFromImpl<T> : 
    Parameters<T> extends [ClientLoaderFunctionArgs | ClientActionFunctionArgs] ?
      Output extends Promise<TypedDeferredData<infer D>> ? D : // client defer() is just the raw data
      Awaited<Output> // client naked object is just the object
    : 
    SerializeFromImpl<T>
  : 
  SerializeFromImpl<T>
);
```